middlewares/sticky_backend_handler.go

```diff
@@ -0,0 +1,274 @@
+	CookieEncryptKey string
+	CookieName       string
+	Rules            []*template.Template
+	UseCookie        bool
+	UseIP            bool
+	UseRules         bool
+}
+
+// NewStickinessParsed creates a new StickinessParsed instance
+func NewStickinessParsed(s *types.Stickiness, backendName string, cookieName string) *StickinessParsed {
+	var useCookie, useIP, useRules bool
+	rules := make([]*template.Template, 0)
+	useDefault := true
+	if len(s.Rules) > 0 {
+		useRules = true
+		useDefault = false
+		for i, rule := range s.Rules {
+			t, err := template.New(guuid.NewV4().String()).Funcs(sprig.FuncMap()).Parse(rule)
+			if err != nil {
+				log.Errorf("Backend %s: failed to parse sticky rule with index %d, error: %s", backendName, i, err)
+				t = nil
+			}
+			rules = append(rules, t)
+		}
+	}
+	if s.IP {
+		useIP = true
+		useDefault = false
+	}
+	if s.Cookie || useDefault {
+		useCookie = true
+	}
+
+	return &StickinessParsed{
+		BackendName:      backendName,
+		CookieEncryptKey: s.CookieEncryptKey,
+		CookieName:       cookieName,
+		Rules:            rules,
+		UseCookie:        useCookie,
+		UseIP:            useIP,
+		UseRules:         useRules,
+	}
+}
+
+// StickyBackendHandler inspects criteria such as Client IP address and uses
+// consistent hashing to route a request to a specific backend server
+type StickyBackendHandler struct {
+	lb   healthcheck.LoadBalancer
+	next http.Handler
+	sp   *StickinessParsed
+}
+
+// NewStickyBackendHandler creates a new StickyBackendHandler instance
+func NewStickyBackendHandler(lb healthcheck.LoadBalancer, next http.Handler, sp *StickinessParsed) *StickyBackendHandler {
+	return &StickyBackendHandler{lb: lb, next: next, sp: sp}
+}
+
+// ServeHTTP inspects the request for a sticky criteria and inserts the
+// sticky header if a criteria is met.  It then invokes the next handler
+// in the middleware chain.
+func (h *StickyBackendHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+
+	var server string
+	var key string
+	requestCookie, requestCookieErr := r.Cookie(h.sp.CookieName)
+	stripSetCookie := true
+
+	if h.sp.UseRules {
+		for i, rule := range h.sp.Rules {
+			if rule != nil {
+				var b bytes.Buffer
+				var w = bufio.NewWriter(&b)
+				err := rule.Execute(w, r)
+				if err != nil {
+					log.Errorf("Backend %s: failed to execute sticky rule with index %d, error: %s", h.sp.BackendName, i, err)
+				}
+				w.Flush()
+				key = strings.TrimSpace(b.String())
+				if key != "" {
+					break
+				}
+			}
+		}
+	}
+
+	if key == "" && h.sp.UseIP {
+		var clientIP string
+		if xForwardedFor := r.Header.Get(forward.XForwardedFor); xForwardedFor != "" {
+			clientIP = strings.TrimSpace(strings.Split(xForwardedFor, ",")[0])
+		} else {
+			clientIP, _, _ = net.SplitHostPort(r.RemoteAddr)
+		}
+		if clientIP != "" {
+			key = "clientIP: " + clientIP
+		}
+	}
+
+	if key != "" {
+		servers := h.lb.Servers()
+		if len(servers) > 0 {
+			sortedServers := make([]string, len(servers))
```

Comment:
Why not pre-sort it once during initialization and store as a struct field, instead of having to sort the same list for every invocation?  Granted that the list is unlikely to be very large, but even for a few dozen entries on a high load endpoints this will cause increase in latency that is non-negligible.

Reply:
I was under the impression that a Load Balancer's servers came from the [Servers() method in oxy/roundrobin/rr.go](https://github.com/vulcand/oxy/blob/856ab519c4cede21de1ecb8cf3507e3d056ce2ce/roundrobin/rr.go#L188-L197)

Since the Load Balancer could take servers in/our of rotation at any time and return them in any order, a fresh list must be retrieved and sorted on each request.

Reply:
Good point.  That’s certainly a possibility.  Was just trying to see if we can optimize it somehow and not have to sort it every proxy request.

Reply:
Just thought about it more and realized that even though Oxy “could” do that dynamically, in practice Traefik would not, as each change to the config from the provider triggers an internal config refresh, which actually recreates the Oxy router with all the Lb and other goodness.  @ldez, @emile, @timoreiman - can you guys confirm that?

Reply:
I think Traefik's `healtcheck.go` calls [UpsertServer](https://github.com/containous/traefik/blob/89d90de7d8f8369a7e30e704d3e88e3ea9b5447e/healthcheck/healthcheck.go#L121) and [RemoveServer](https://github.com/containous/traefik/blob/89d90de7d8f8369a7e30e704d3e88e3ea9b5447e/healthcheck/healthcheck.go#L132), so the result of `Servers()` is still changing throughout the lifetime of a single configuration and it's result could vary request-to-request